facebook/ThreatExchange
# [py-tx] Big ol' Data Model and experimental-fetch Rewrite

Summary
---------
I put up this PR after a few days of rewriting this over and over again. To try and finish it off, I'm forcing myself to commit what I consider a workable version before I continue on this path.

This diff does a few interesting things:
1. Recycles most of the previous implementation of experimental-fetch
2. Attempts a partial rewrite of the command to make it more similar to fetch
3. Defines the storage layer for threat_updates, which is basically to store a minified version of the raw json from the endpoint

My default plan is then to do:
```
Sync -> Raw json output
 => SimpleDescriptorRollup (since it works on tag fetch endpoint too) - or a variation thereof

HMA:
  => SimpleDescriptorRollup -> DB
    => SignalType marks which rollups belong to a signal type
       => Output file that is simple Hash -> DB_ids (many options here)
          => SignalType.buildIndex(ofile) -> Index
             => Index goes to matcher
               => SignalType.hash(content) -> Hash -> Index.match(Hash) -> {db_id: dist}
                  => Matcher fetched from DB to determine what to do

CLI:
fetch -> stores minified raw json
  => SimpleDescriptorRollup
    => SignalType marks which rollups belong to a signal type
       => Store indexes for signal type
match
  => determine SignalType from content type
    => load indexes
      => most indexes will be linear scan SignalType.hash(content) -> SignalType.match(content, stored_hash)
        => output match results from SimpleDescriptorRollup
```
However, much of the code is in placeholder locations, and not everything is connected.


Test Plan
---------
tl;dr:
1. Run fetch
2. Run experimental-fetch
See that the files are the same size (though not quite same content)

```
$  time threatexchange -c ~/test.te.cfg fetch

$  time threatexchange -c ~/test.te.cfg experimental-fetch
Currently at ages long past
Downloaded 501 updates. Currently at 67 days 17h16m52s ago
^C
Exception occurred! Attempting to save...

real    0m7.159s

# Resumes
$  time threatexchange -c ~/test.te.cfg experimental-fetch
Currently at 67 days 17h14m25s ago
Downloaded 1001 updates. Currently at 67 days 17h12m03s ago
...

$ ls ~/output_dir
total 900K
-rw-r--r-- 1 dcallies users   51 Feb 17 13:50 7.....threat_updates.te
-rw-r--r-- 1 dcallies users 330K Feb 16 18:32 DEBUG_STRING.te
-rw-r--r-- 1 dcallies users 132K Feb 16 18:32 HASH_PDQ_OCR.te
-rw-r--r-- 1 dcallies users  77K Feb 16 18:32 HASH_PDQ.te
-rw-r--r-- 1 dcallies users 344K Feb 16 18:32 HASH_TMK.te
-rw-r--r-- 1 dcallies users  750 Feb 16 18:32 RAW_URI.te

$ threatexchange -c ~/test.te.cfg fetch
# Doesn't asplode

$ ls # Edited with vim
  69 531732984141635.threat_updates.checkpoint
  37 fetch_checkpoint.te
132M raw_text.te
132M simple.TEXT_STRING.te
133M simple.DEBUG_STRING.te
 54K video_md5.te
 60K simple.HASH_MD5.te
 260 simple.HASH_VIDEO_MD5.te
 54K pdq_ocr.te
 56K simple.HASH_PDQ_OCR.te
 85K pdq.te
 89K simple.HASH_PDQ.te
 66M video_tmk_pdqf.te
 66M simple.HASH_TMK.te
 80K trend_query.te
5.1K simple.TREND_QUERY.te  # Backfill hasn't completed
3.5M simple.URI.te  # Regular URL has a bug!
```
